R/meltAssay.R

```diff
@@ -0,0 +1,135 @@
+  function(x, colData_choice = "all", abund_values = "counts") {
+    # input check
+    if (isFALSE(any(abund_values == c("counts", "relabundance", "both")))) {
+      stop(paste0(
+        "'abund_values' must be one:  ",
+        list(assayNames(x))
+      ), " or", " 'both' ")
+    }
+
+    uTaxaID <- uSamId <- Abundance <- NULL
+
+    if (is.null(rowData(x))) {
+      warning("'rowData(x)' is NULL!")
+    } else {
+      # Get taxonomic data stored in rowData
+      me_tax <- rowData(x) %>%
+        data.frame() %>%
+        # uTaxaID is unique TaxaID
+        rownames_to_column("uTaxaID")
+      # may need to enforce in case of user supplied duplicate
+    }
+
+    if (is.null(assay(x))) {
+      stop("'rowData(x)' is NULL!")
+    }
+
+    if (abund_values == "counts") {
+      dfSE <- assays(x)$counts %>%
+        data.frame() %>%
+        rownames_to_column("uTaxaID") %>%
+        # uSamId is unique sample id
+        pivot_longer(!uTaxaID,
+          values_to = "Abundance",
+          names_to = "uSamId"
+        )
+    } else if (abund_values == "relabundance") {
+      dfSE <- assays(x)$relabundance %>%
+        data.frame() %>%
+        rownames_to_column("uTaxaID") %>%
+        pivot_longer(!uTaxaID,
+          values_to = "Abundance",
+          names_to = "uSamId"
+        )
+    } else if (abund_values == "both") {
+      me_count_ab <- assays(x)$counts %>%
+        data.frame() %>%
+        rownames_to_column("uTaxaID") %>%
+        pivot_longer(!uTaxaID,
+          values_to = "CountAbundance",
+          names_to = "uSamId"
+        )
+
+      me_rel_ab <- assays(x)$relabundance %>%
+        data.frame() %>%
+        rownames_to_column("uTaxaID") %>%
+        pivot_longer(!uTaxaID,
+          values_to = "RelAbundance",
+          names_to = "uSamId"
+        )
+
+      dfSE <- me_count_ab %>%
+        left_join(me_rel_ab, by = c("uTaxaID", "uSamId"))
+    }
+
+    if (is.null(rowData(x)) & is.null(colData_choice) | is.na(colData_choice)) {
+      return(dfSE)
+    } else if (!is.null(rowData(x)) & is.null(colData_choice) | is.na(colData_choice)) {
+      dfSE <- dfSE %>%
+        left_join(me_tax, by = "uTaxaID") %>%
+        mutate_if(is.factor, as.character)
+
+      return(dfSE)
+    }
+
+    # check coldata
+    if (colData_choice == "all") {
+      me_sam <- colData(x) %>%
+        data.frame() %>%
+        rownames_to_column("uSamId")
+
+      dfSE <- dfSE %>%
+        left_join(me_sam, by = "uSamId") %>%
+        mutate_if(is.factor, as.character)
+
+      return(dfSE)
+    } else if (isTRUE(any(colData_choice == colnames(colData(x))))) {
+      colData_choice.n <- intersect(colData_choice, colnames(colData(x)))
+      sub.colData <- colData(x)[, colData_choice.n]
+
+      me_sam <- sub.colData %>%
+        data.frame() %>%
+        rownames_to_column("uSamId")
+
+      dfSE <- dfSE %>%
+        left_join(me_sam, by = "uSamId") %>%
+        mutate_if(is.factor, as.character)
+
+      return(dfSE)
+    }
```

Comment:
I don't see the necessity for distinguishing `all` and `some`. technically is doesn't make a difference, which is shown by the fact that lines 115-117 and 128-130 are duplicates.

I would suggest, selecting the colData to use based on the chosen colums and then `left_join` them independent on how many there are.

In addition, one usage case is missing, which is `none`. I would argue, this should also be the default defined by the value `NULL`

Furthermore, `colData_choice` is mixing snake and camel case. I would strongly advise to revised this. Maybe `add_sample_data` or `add_col_data` would be a better name. Then the whole function could be formulated as a sentence, which is sometimes a good sanity check for naming arguments. The sentence `melt the assay with the abund_values "counts" and add the sample data "x,y and z" ` works. With colData_choice the sentence gets kind of clunky.